### PR TITLE
Try to split physical from logical

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

While doing the previous PRs and at other times I've noticed there is unnecessary coupling between `QueryMetadata` to the logical query model, (basically `QueryMetadata` has a `outputNode` field that points to the logical query model), and that the line between the logical and physical query building / model is often blurred.

This makes it harder for people to reason about what the code is doing.  

In this PR I'm looking to improve things slightly.  There are two related pieces of work in this PR, in two separate commits:

- commit one drops the `OutputNode` field from the `QueryMetadata`.  Once built, the query shouldn't need to reference the logical query model used to build it.  Splitting them decouples these areas of code and makes the model available for garbage collection. 
- commit two splits the `QueuedSchemaKStream`, which is created as the SINK node in the logical query model, but also contains physical details like the blocking queue and the thing that will actually write to the queue.  The physical part has been copied out into `TransientQueryQueue`.

### Testing done 
The usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.physical;
+
+/**
+ * @author andy created 2019-03-13
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java]: Looks a little different from your usual interface java docs...

File: ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java [hunk 2]
```diff
@@ -18,41 +18,48 @@
...
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
 import java.util.function.Consumer;
+import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.streams.KafkaStreams;
 import org.apache.kafka.streams.Topology;
 
+/**
+ * Metadata of a persistent query, e.g. {@code SELECT CREATE STREAM FOO AS SELECT * FROM BAR;}.
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java]:
```suggestion
 * Metadata of a persistent query, e.g. {@code CREATE STREAM FOO AS SELECT * FROM BAR;}.
```

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/entity/QueryDescriptionTest.java [hunk 3]
```diff
@@ -15,173 +15,136 @@
...
           .field("field1", SchemaBuilder.int32().build())
           .field("field2", SchemaBuilder.string().build())
           .build();
+
+  private static final List<FieldInfo> EXPECTED_FIELDS = Arrays.asList(
+      new FieldInfo("field1", new SchemaInfo(Type.INTEGER, null, null)),
+      new FieldInfo("field2", new SchemaInfo(Type.STRING, null, null)));
+
   private static final String STATEMENT = "statement";
+  private static final Map<String, Object> STREAM_PROPS = Collections.singletonMap("k1", "v1");
```
Comment (vcrfxia) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/entity/QueryDescriptionTest.java]:
```suggestion
  private static final Map<String, Object> STREAMS_PROPS = Collections.singletonMap("k1", "v1");
```
To be consistent with our naming elsewhere.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java [hunk 4]
```diff
@@ -119,59 +121,62 @@ public QueryMetadata buildPhysicalPlan(final LogicalPlanNode logicalPlanNode) {
...
       );
+    }
 
-    } else if (outputNode instanceof KsqlStructuredDataOutputNode) {
+    if (outputNode instanceof KsqlStructuredDataOutputNode) {
+      if (resultStream instanceof QueuedSchemaKStream) {
+        throw new KsqlException(String.format(
+            "Mismatch between logical and physical output; "
+                + "expected a SchemaKStream based on logical "
+                + "KsqlBareOutputNode, found a %s instead",
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java]:
```suggestion
                + "KsqlStructuredDataOutputNode, found a %s instead",
```
Typo.

File: ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java [hunk 5]
```diff
@@ -199,9 +208,45 @@ public void shouldHaveKStreamDataSource() {
...
+  @Test
+  public void shouldBuildMapValuesNodeForTransientQueries() {
+    final QueryMetadata query = buildPhysicalPlan(simpleSelectFilter);
+
+    final TopologyDescription.Processor node = getNodeByName(query, FILTER_MAPVALUES_NODE);
+
+    verifyProcessorNode(node, ImmutableList.of(FILTER_NODE), ImmutableList.of(FOREACH_NODE));
+  }
+
+  private static Processor getNodeByName(final QueryMetadata query, final String nodeName) {
```
Comment (vcrfxia) [inline ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java]: nit: move to bottom of file. Also, no given/when/then pattern for these new tests?

File: ksql-engine/src/main/java/io/confluent/ksql/physical/QueueCallback.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+package io.confluent.ksql.physical;
+
+/**
+ * A callback called before and after things are queued.
+ */
+public interface QueueCallback {
+
+  /**
+   * Called to determine is an output row should be queued for output.
```
Comment (vcrfxia) [inline ksql-engine/src/main/java/io/confluent/ksql/physical/QueueCallback.java]:
```suggestion
   * Called to determine if an output row should be queued for output.
```
nit: typo.

File: ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java [hunk 7]
```diff
@@ -150,15 +150,17 @@ public void closeEngine() {
 
   @Test
   public void shouldCreatePersistentQueries() {
+    // When"
```
Comment (vcrfxia) [inline ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java]:
```suggestion
    // When:
```
super nit.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlBareOutputNodeTest.java [hunk 8]
```diff
@@ -102,18 +102,6 @@ public void shouldBuildFilterNode() {
     verifyProcessorNode(node, Collections.singletonList(TRANSFORM_NODE), Collections.singletonList(FILTER_MAPVALUES_NODE));
   }
 
-  @Test
-  public void shouldBuildMapValuesNode() {
-    final TopologyDescription.Processor node = (TopologyDescription.Processor) getNodeByName(FILTER_MAPVALUES_NODE);
-    verifyProcessorNode(node, Collections.singletonList(FILTER_NODE), Collections.singletonList(FOREACH_NODE));
```
Comment (vcrfxia) [inline ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlBareOutputNodeTest.java]: `FOREACH_NODE` is now unused and can be removed from the top of this file.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java]: lol. Ooops.

File: ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java]: and again... oops! Must be working too hard.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java]: done.

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/entity/QueryDescriptionTest.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/test/java/io/confluent/ksql/rest/entity/QueryDescriptionTest.java]: done.

File: ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java]: Terribly sorry. They were a cut & paste job from another file.  Sorted :D

File: ksql-engine/src/main/java/io/confluent/ksql/physical/QueueCallback.java (hunk 6 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/QueueCallback.java]: done.

File: ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlBareOutputNodeTest.java (hunk 8 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlBareOutputNodeTest.java]: done.

File: ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java (hunk 7 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/engine/KsqlEngineTest.java]: they're the best type.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/LimitedQueueCallback.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  @Override
+  public void setLimitHandler(final LimitHandler limitHandler) {
+    this.limitHandler = Objects.requireNonNull(limitHandler, "limitHandler");
+  }
+
+  @Override
+  public boolean shouldQueue() {
+    return queued.get() > 0;
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/LimitedQueueCallback.java]: Good catch - I'd not spotted / remembered this was accessed from several threads.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/LimitHandler.java]: Already removed.

File: ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/util/PersistentQueryMetadata.java]: Already removed.

File: ksql-engine/src/main/java/io/confluent/ksql/physical/TransientQueryQueue.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * A queue of rows for transient queries.
+ *
+ * @param <K> the key type.
+ */
+class TransientQueryQueue<K> {
+
+  private final LimitQueueCallback callback;
+  private final BlockingQueue<KeyValue<String, GenericRow>> rowQueue =
+      new LinkedBlockingQueue<>(100);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/TransientQueryQueue.java]: Not in this PR - this is just code moved out of `QueuedSchemaKStream`

File: ksql-engine/src/main/java/io/confluent/ksql/physical/TransientQueryQueue.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+  TransientQueryQueue(final QueuedSchemaKStream<K> schemaKStream) {
+    final OutputNode output = schemaKStream.outputNode();
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/physical/TransientQueryQueue.java]: The output node is being used here to _build_ the physical plan.  Which must be the case...

What's different is that we're not storing a reference to the output node in the physical plan / query metadata any more.

**Pull request merged**